noir_stdlib/src/collections/map.nr

```diff
@@ -0,0 +1,305 @@
+use crate::cmp::Eq;
+use crate::collections::vec::Vec;
+use crate::option::Option;
+use crate::default::Default;
+use crate::hash::{Hash, Hasher, BuildHasher};
+
+// Generic trait for one-directional key -> value data structure.
+// May be implemented by either hash tables or graphs for instance.
+trait Map<K, V> {
+    fn len(self) -> u64;
+    fn get(self, key: K) -> Option<V>;
+    fn insert(&mut self, key: K, value: V);
+    fn remove(&mut self, key: K);
+}
+
+// Hash table with open addressing and quadratic probing.
+// Size of the underlying table must be known at compile time.
+// It is advised to select capacity N as a power of two, or a prime number 
+// because utilized probing scheme is best tailored for it.
+struct HashMap<K, V, N, B> {
+    _table: [Slot<K, V>; N],
+
+    // Amount of valid elements in the map.
+    _len: u64,
+
+    _build_hasher: B
+}
+
+// Data unit in the HashMap table.
+// In case Noir would support enums in future, this  
+// better to be refactored to use it with three states:
```

Comment:
```suggestion
// Data unit in the HashMap table.
// In case Noir adds support for enums in the future, this  
// should be refactored to have three states:
```